### ISSUE TITLE
No alpha_ attribute in SpaceNetClassifier decoder

### ISSUE THREAD

**Issue by vukovicnikola** [nilearn/nilearn#1566]
If I run a simple fit using, for example: 
```
cv = StratifiedKFold(n_folds=2,y=y)

decoder = SpaceNetClassifier(cv=cv, memory="nilearn_cache", penalty='tv-l1', memory_level=2, n_jobs=2, verbose=1, screening_percentile=20, mask= './mni_152_gm.nii')
decoder.fit(X, y)
```

According to the [Reference](http://nilearn.github.io/modules/generated/nilearn.decoding.SpaceNetClassifier.html),  there should be a decoder.alpha_ attribute. However, this does not exist when I inspect the decoder.

How to get the decoding accuracy without the best alpha from cv? Were there changes to SpaceNetClassifier since the reference page was last updated?

**Comment by bthirion** [comment]
Sorry, I don't see the issue. Can you reproduce it on one of the examples provided with nilearn, e.g. 
http://nilearn.github.io/auto_examples/02_decoding/plot_mixed_gambles_space_net.html#sphx-glr-auto-examples-02-decoding-plot-mixed-gambles-space-net-py ?

**Comment by MadsJensen** [comment]
@bthirion Hi just ran the example you mentioned above (just downloaded the .py and ran it). I get "infs" in the "decoder.cv_scores"  and I have no "decoder.alpha_" either. 

Python = 3.5
nilearn = 0.4
scikit-learn = 0.19.0

**Comment by vukovicnikola** [comment]
Yup. Same here. 

```
decoder.cv_scores_
Out[2]: 
array([[[       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.33628491,  0.33951149,  0.35399992,  0.35705147],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.22993435,  0.1948318 ,  0.18355382,  0.14858419],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.02179566,  0.02130482,  0.02092132,  0.0257029 ],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.15211956,  0.13849411,  0.13876513,  0.13249187],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf, -0.05248152, -0.04071703, -0.02701244, -0.01560909],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.00977761,  0.01555971,  0.03138973,  0.03321216],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.24941553,  0.28438571,  0.28277637,  0.27943689],
        [       -inf,        -inf,        -inf,        -inf,        -inf,
                -inf,  0.20014558,  0.20458997,  0.21957429,  0.23424287]]])
```

I should note however that the output coef image looks fine. 
![img](https://user-images.githubusercontent.com/6719774/33035928-21efa50a-ce2d-11e7-8580-20ffccb2bfd2.png)

**Comment by KamalakerDadi** [comment]
Sorry, for being so late in updating SpaceNet documentation. The documentation needs to be updated. `alpha_` is not an attribute.

The attribute you are interested in is
 `best_model_params_` : ndarray, shape (n_folds, n_parameter)
        Best model parameters (alpha, l1_ratio) saved for the different
        cross-validation folds.

**Comment by MadsJensen** [comment]
@KamalakerDadi  Great thanks for the clarification. But what about the "inf" in the cv_scores_ ? Is that expected? (Sorry I don't know the details of the SpaceNet algorithm.)

**Comment by vukovicnikola** [comment]
And a final question - is there an out of the box method to get the mean prediction accuracy (taking into account the best alphas + these strange -inf scores)? Thanks!

**Comment by KamalakerDadi** [comment]
>But what about the "inf" in the cv_scores_ ? Is that expected? (Sorry I don't know the details of the SpaceNet algorithm.)

I don't know the algorithm very well too. But, pointing towards to look more closely into it. I can update your question soon.

**Comment by bthirion** [comment]
Really weird. Can you reproduce on the datasets provided as examples ?
Are there some cross-validation folds with constant data ?

**Comment by vukovicnikola** [comment]
The above outputs *are* from example data. I also get these -inf values in cv_scores_ when I use my own data, and with different number of Kfolds...

**Comment by bthirion** [comment]
Hm. T am realizing that plot_haxby_space_net.py is broken. Yet, this was not detected by CircleCI...
Don't understand what's going on here.

**Comment by MadsJensen** [comment]
ok, let's know if there is anything you would like from our side (@vukovicnikola or me) .

**Comment by bthirion** [comment]
OK. I had to upgrade some of the libraries to make it work ;-(
@vukovicnikola I reproduce your result, but this is fine: for some values of the hyperparameters, the optimization fails, yielding a score of -\infty. This would be worrisome if we had that for the whole grid, but this is not the case. It simply reminds us that these beasts are impossible to deal with when hyperparameters are poorly set (e.g. not enough regularization).
In detail, somebody whould check whether -\infty comes from numerical divergence --which should not occur-- or is a default value in the grid

**Comment by KamalakerDadi** [comment]
>In detail, somebody whould check whether -\infty comes from numerical divergence --which should not occur-- or is a default value in the grid

Values are coming from while computing test score for model given weight maps 'w'. This line
https://github.com/nilearn/nilearn/blob/e34697bc22ee0061bbb548c89246cd814f23153a/nilearn/decoding/space_net.py#L259. I guess values are in the grid

**Comment by bthirion** [comment]
So, I see nothing abnormal here.

**Comment by KamalakerDadi** [comment]
Yes, except that the some of the documentation should be fixed which I will do it soon.

**Comment by vukovicnikola** [comment]
Thank you @KamalakerDadi and @bthirion for your input - it's great to hear that the -inf values are not due to some more fundamental problem, which would impact the overall analysis. I will then just write some code to bypass these and calculate mean model accuracy based on the best alphas from .best_model_params_. Thanks @KamalakerDadi for dealing with the Docs - I'm sure it will be helpful to other users as well...
